internal/controller/grpc_logger.go

```diff
@@ -116,17 +119,36 @@ func (b *BufferedGrpcWriteSyncer) flush() {
 	}
 }
 
+// sendLogsKeepalive sends a keepalive ping on the logs stream
+func (b *BufferedGrpcWriteSyncer) sendLogsKeepalive() error {
+	return b.client.Send(&pb.SendLogsRequest{
+		Request: &pb.SendLogsRequest_Keepalive{
+			Keepalive: &pb.Keepalive{},
+		},
+	})
+}
+
 // run flushes the buffer at the configured interval until Stop is called.
 func (b *BufferedGrpcWriteSyncer) run() {
 	ticker := time.NewTicker(logFlushInterval)
 	defer ticker.Stop()
-
+	keepAliveTicker := time.NewTicker(jitterTime(keepAlivePeriod, 0.10))
+	defer keepAliveTicker.Stop()
 	for {
 		select {
 		case <-ticker.C:
 			b.mutex.Lock()
 			b.flush()
 			b.mutex.Unlock()
+		case <-keepAliveTicker.C:
+			b.mutex.Lock()
```

Comment:
Locking here is a good call.
In the gRPC logger, any goroutine may write a log message, which may result in a Send call to the stream.
So all usages of the stream must be protected by the mutex.